components/plan-build/bin/environment.bash

```diff
@@ -204,80 +212,61 @@ __populate_environment_from_deps() {
-# Don't need to do this for packages that already have a
-# RUNTIME_ENVIRONMENT file, because this operation will have already
-# been done when those packages were built. This is to help preserve
-# backward compatibility.
-__assemble_legacy_path() {
-    local path_to_dep="${1}"
-    local path=""
-
-    local tdeps
-    tdeps=$(_read_metadata_file_for ${path_to_dep} "TDEPS")
-    if [[ -n "${tdeps}" ]]; then
-        for tdep in ${tdeps}; do
-            local path_to_tdep="${HAB_PKG_PATH}/${tdep}"
-            # If a PATH exists for this transitive dependency, collect
-            # it, prepending to the beginning of our accumulating path
-            local tdep_path
-            tdep_path=$(_read_metadata_file_for "${path_to_tdep}" "PATH")
-            if [[ -n "${tdep_path}" ]]; then
-                path=$(push_to_path "${tdep_path}" "${path}")
-            fi
-        done
-    fi
-    # If the current dependency has PATH data itself, prepend it to
-    # what we've been accumulating so far
-    local current_dep_path
-    current_dep_path=$(_read_metadata_file_for "${path_to_dep}" "PATH")
-    if [[ -n "${current_dep_path}" ]]; then
-        path=$(push_to_path "${current_dep_path}" "${path}")
+# The path is constructed by taking all `PATH` metadata file entries from this
+# package (in for the form of `$pkg_bin_dirs[@]`), followed by entries from the
+# *direct* dependencies first (in declared order), and then from any remaining
+# transitive dependencies last (in lexically sorted order). All entries are
+# present only once in the order of their first appearance.
+__assemble_runtime_path() {
+  local paths=()
+  local dir dep data
+
+  # Add element for each entry in `$pkg_bin_dirs[@]` first
+  for dir in "${pkg_bin_dirs[@]}"; do
+    paths+=("$pkg_prefix/$dir")
+  done
+
+  # Iterate through all direct direct run dependencies following by all
+  # remaining transitive run dependencies and for each, append each path entry
+  # onto the result, assuming it hasn't already been added. In this way, all
+  # direct dependencies will match first and any programs that are used by a
+  # direct dependency will also be present on PATH, albeit at the very end of
+  # the PATH.
+  for dep in "${pkg_deps_resolved[@]}" "${pkg_tdeps_resolved[@]}"; do
+    if [[ -f "$dep/PATH" ]]; then
+      data="$(cat "$dep/PATH")"
+      data="$(trim "$data")"
+      while read -r entry; do
+        paths=($(_return_or_append_to_set "$entry" "${paths[@]}"))
+      done <<< $(echo "$data" | tr ':' '\n')
     fi
+  done
+
+  # Return the elements of the result, joined with a colon
+  join_by ':' "${paths[@]}"
+}
 
-    # Finally, return the path
-    echo "$path"
+# **Internal**  Set the `PATH` key for the runtime environment if a computed
+# runtime path is necessary. If a package has no `${pkg_bin_dirs[@]}` elements
+# and has no runtime dependencies with `PATH` entries, then a computed runtime
+# path would be empty, meaning that no `PATH` value should be set.
```

Comment:
Might be worth noting that we are only doing this to ensure that any services built with this version of the build program can still run in Supervisors between when `RUNTIME_ENVIRONMENT` was introduced and now, which de facto assume the presence of a `PATH` variable (and any non-services that you might run via `hab pkg exec`, I suppose). Supervisors after this time will use a separate `RUNTIME_PATH` file which will get layered on top (spoiler alert for those reading along at home 😜 )

Reply:
I was planning on changing this comment with the next commit to make this change.

Oh, and I forgot to call this out specially, but I dropped the provenance of `PATH` for `RUNTIME_ENVIRONMENT_PROVENANCE` as a first step to making this a backwards compatible change.

Reply:
Addressed now.